### ISSUE TITLE
server,storage: fail hard when a store is missing

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#69819]
**Describe the problem**

We found this situation in a [customer issue](https://github.com/cockroachlabs/support/issues/1182): customer started a node with a 2nd `--store` flag, then stopped the node, then restarted the node without the `--store` flag and more importantly **after marking the underlying data directory as unreachable** (I presume they also deleted the data).

The result is that (duh) the 2nd store fails to serve requests
```
I210823 14:26:48.055505 2680 kv/kvserver/store.go:1926 ⋮ [n2,s2] 16  [n2,s2]: failed initial metrics computation: open ‹/opt/cockroach-data/tmp›: permission denied
```

However, the real problem is that **the store still initializes without error**. As a result, the node starts up successfully and **no clear indication is given to the operator that they have created a broken node**.

**To Reproduce**

1. add extra store
2. remove extra store flag and corresponding data dir
3. restart node

**Expected behavior**

The server process fails to start entirely with a clear exit error. 

**Additional data / screenshots**

Symptoms of the missing store directory:

- node appears to serve requests (as long as no ranges in the missing store are requested)
- `crdb_internal.kv_store_status` missing a row for the node
- cluster UI reports node capacity as "NaN%"
- log errors:
  ```
  I210823 14:26:48.055505 2680 kv/kvserver/store.go:1926 ⋮ [n2,s2] 16  [n2,s2]: failed initial metrics computation: open ‹/opt/cockroach-data/tmp›: permission denied
  E210823 14:26:48.070450 67 server/status/recorder.go:469 ⋮ [n2,summaries] 24  could not record status summaries: Store 2 could not return descriptor, error: open ‹/opt/cockroach-data/tmp›: permission denied
  ```

cc @jbowens for triage

**Comment by knz** [comment]
This may be a KV issue as well.

- given the lack of range operation errors in the linked issues, I am assuming the extra store did not actually contain any range. So *in theory* cluster stability is not impaired.
- however, we do need to compute metrics + status summaries for all stores on every node.

**Arguably, we should be able to ignore store errors when computing metrics / status summaries and store a partial node status record.**

@tbg what do you think?

**Comment by tbg** [comment]
Could the user provide a more direct repro of this? The store directory contains the store identifier and if CockroachDB is started without being pointed at a given store, it will not attempt to touch that store. I verified this using

> ./cockroach start --store store1 --store store2 --store store3 --insecure --logtostderr --join 127.0.0.1:26257

and init'ing that node, then stopping it after a short while and restarting without the `--store store3` flag. The node worked normally and no error messages about s3 were logged.

From the symptoms above, it seems as though the user had made `store3` non-writeable but retained the store flag?

**Comment by knz** [comment]
> From the symptoms above, it seems as though the user had made store3 non-writeable but retained the store flag?

We have the log files. There is just one store flag (` --store=/opt/cockroach-data`)

**Comment by knz** [comment]
What about the node started up with a 2nd `--store` flag *before it joined the cluster* (and thus before it got its node/store  IDs) and then restarted without it?

**Comment by tbg** [comment]
No, that doesn't make a difference. The source of truth of which stores you have are the stores passed via the stores flags, from which the StoreIdent is then loaded.

> We have the log files. There is just one store flag ( --store=/opt/cockroach-data)

Ok, and the error says

> [n2,s2] 16  [n2,s2]: failed initial metrics computation: open ‹/opt/cockroach-data/tmp›: permission denied

so isn't this just telling us that someone messed around with the permissions of /opt/cockroach-data (and that this directory is s2)?

**Comment by knz** [comment]
let's ask the customer for a listing.
